src/tui/terminal.rs

```diff
@@ -0,0 +1,321 @@
+use crate::tui::{
+    backend::Backend,
+    buffer::Buffer,
+    layout::Rect,
+    widgets::{StatefulWidget, Widget},
+};
+use std::io;
+
```

Comment:
Hi @conradludgate @ellie,

I started to try and rebase my crossterm/inline PR https://github.com/ellie/atuin/pull/648, and figured that the version of tui-rs that was copied here is missing the bits needed from https://github.com/fdehau/tui-rs/pull/552/files#diff-554109cedfab40c7397c98d70e7a7449266358293b29ad5aa8db2f79e958122bR11 to make inline viewport possible.

Should I try and backport https://github.com/fdehau/tui-rs/pull/552 on this local copy?

Any reason why git history from upstream was not preserved using git subtree?

Also, wouldn't maintaining a fork of tui-rs make things easier, at least until it is not needed anymore at all?

Reply:
A full fork WG is already in the works and I'm taking part as a stakeholder. It seems we won't be vendoring it for long.

I'm not an expert in all the git features so I didn't think to use subtree. It sounds like it would be awkward though because we want it as a module and not a crate

Reply:
Ok, I guess should I copy the parts needed from the upstream inline PR then.